validator/beacon/service.go

```diff
@@ -61,68 +87,95 @@ func (s *Service) Stop() error {
-	stream, err := client.LatestBeaconBlock(s.ctx, &empty.Empty{})
-	if err != nil {
-		log.Errorf("Could not setup beacon chain block streaming client: %v", err)
-		return
+	crystallized := res.GetLatestCrystallizedState()
+	if err := s.processCrystallizedState(crystallized); err != nil {
+		log.Fatalf("unable to process received crystallized state: %v", err)
 	}
-	for {
-		block, err := stream.Recv()
+}
 
-		// If the stream is closed, we stop the loop.
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			log.Errorf("Could not receive latest beacon block from stream: %v", err)
-			continue
-		}
-		log.WithField("slotNumber", block.GetSlotNumber()).Info("Latest beacon block slot number")
-
-		// Based on the slot determined from the latest crystallized state, check if
-		// it matches the latest received beacon slot.
-		if s.responsibility == "proposer" {
-			log.WithField("slotNumber", block.GetSlotNumber()).Info("Assigned proposal slot number reached")
-			s.responsibility = ""
-			s.proposerAssignmentFeed.Send(block)
-		} else if s.responsibility == "attester" && block.GetSlotNumber() == s.assignedSlot {
-			// TODO: Let the validator know a few slots in advance if its attestation slot is coming up
-			log.Info("Assigned attestation slot number reached")
-			s.responsibility = ""
-			s.attesterAssignmentFeed.Send(block)
+// waitForAssignment kicks off once the validator determines the currentSlot of the
+// beacon node by calculating the difference between the current system time
+// and the genesis timestamp. It runs exactly every SLOT_LENGTH seconds
+// and checks if it is time for the validator to act as a proposer or attester.
+func (s *Service) waitForAssignment(ticker <-chan time.Time, client pb.BeaconServiceClient) {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker:
+			log.WithField("slotNumber", s.currentSlot).Info("New beacon node slot interval")
+			if s.responsibility == "proposer" && s.assignedSlot == s.currentSlot {
+				log.WithField("slotNumber", s.currentSlot).Info("Assigned proposal slot number reached")
+				s.responsibility = ""
+				block, err := client.CanonicalHead(s.ctx, &empty.Empty{})
+				if err != nil {
+					log.Errorf("Could not fetch canonical head via gRPC from beacon node: %v", err)
+					continue
+				}
+				// We forward the latest canonical block to the proposer service via a feed.
+				s.proposerAssignmentFeed.Send(block)
+			} else if s.responsibility == "attester" && s.assignedSlot == s.currentSlot {
+				log.Info("Assigned attestation slot number reached")
+				s.responsibility = ""
+				block, err := client.CanonicalHead(s.ctx, &empty.Empty{})
+				if err != nil {
+					log.Errorf("Could not fetch canonical head via gRPC from beacon node: %v", err)
+					continue
+				}
+				// We forward the latest canonical block to the attester service a feed.
+				s.attesterAssignmentFeed.Send(block)
+			}
+			// Increase the current slot by one every 8 seconds.
+			s.currentSlot++
```

Comment:
do we need to keep track of the current slot when we know that it's `timeSinceGenesis / params.SlotLength` ?

Reply:
hm it makes things easier as I only have to do the timeSinceGenesis / params.SlotLength once during initialization and then I can just increase it this way. It will also be accessed by other methods in the service in a few PRs I have coming up.

Reply:
OK. I'm just worried that if the clock drifts away then is corrected, then this value may not be corrected. 

Another issue is that if the for loop calls `continue` then this is not incremented.
I think this will cause a nasty bug. Can we change this to a getter method that computes the current slot based on the clock time?

Reply:
Let's do that in a separate PR that fixes all the clock issues

Reply:
It'll be a PR that addresses #511 and we can fix all of these slot related issues there